com.unity.multiplayer.mlapi/Runtime/Timing/NetworkTimeSystem.cs

```diff
@@ -0,0 +1,111 @@
+using System;
+using UnityEngine;
+
+namespace MLAPI.Timing
+{
+    /// <summary>
+    /// <see cref="NetworkTimeSystem"/> is a standalone system which can be used to run a network time simulation.
+    /// The network time system maintains both a local and a server time. The local time is based on
+    /// </summary>
+    public class NetworkTimeSystem
+    {
+        private double m_Time;
+
+        private double m_CurrentLocalTimeOffset;
+
+        private double m_DesiredLocalTimeOffset;
+
+        private double m_CurrentServerTimeOffset;
+
+        private double m_DesiredServerTimeOffset;
+
+        /// <summary>
+        /// Gets or sets the amount of time in seconds the server should buffer incoming client messages.
+        /// This increases the difference between local and server time so that messages arrive earlier on the server.
+        /// </summary>
+        public double LocalBuffer { get; set; }
+
+        /// <summary>
+        /// Gets or sets the amount of the time in seconds the client should buffer incoming messages from the server. This increases server time.
+        /// A higher value increases latency but makes the game look more smooth in bad networking conditions.
+        /// </summary>
+        public double ServerBuffer { get; set; }
+
+        /// <summary>
+        /// Gets or sets a threshold in seconds used to force a hard catchup of network time.
+        /// </summary>
+        public double HardResetThreshold { get; set; }
+
+        /// <summary>
+        /// Gets or sets the ratio at which the NetworkTimeSystem speeds up or slows down time.
+        /// </summary>
+        public double AdjustmentRatio { get; set; }
+
+        public double LocalTime => m_Time + m_CurrentLocalTimeOffset;
+
+        public double ServerTime => m_Time + m_CurrentServerTimeOffset;
+
+        //TODO This is used as a workaround to pass this back into the sync function will be removed once we get correct value.
```

Comment:
do you mind elaborating a little bit more? is there a chance to get rid of this workaround completely in the future?

Reply:
Yes this is here because of the lack of the sequence-ack protocol which will be implemented with snapshots. It will be removed hopefully before the next release.

I added a comment which elaborates a bit more:
> Once we have a sequence-ack algorithm we get access to the last received tick time instead.
> See NetworkManager.OnNetworkPreUpdate where this is used. Because we don't receive time sync messages from the server yet we just manually fake a time advancement by adding delta time to our existing time.